### PR TITLE
Clean progress bar code and fix TextureMap patch

### PR DESCRIPTION
**Pull request by mezz**
1. Adds a "test" side to `SidedProvider` to let the progress bars run in testing environments.
2. Cleans up the progress bar code a bit to make it a bit easier to use and encourage good code.
   * Has a new `run` method which lets you wrap your code in a lambda instead of doing push/pop
3. Cleans up the `TextureMap` patch, it was becoming a mess.
   * Fixed a porting bug where we checked if a `String` was contained in the `sprites` map which now takes `ResourceLocation` as keys.
   * Fixed a porting bug where we try to get a `String` from `mapUploadedSprites` which now takes `ResourceLocation` keys.
   * These bugs probably impacted dependent textures like for the bucket, I'll have to look into that later.

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/SidedProvider.java [hunk 1]
```diff
@@ -72,6 +83,9 @@ public static void setServer(Supplier<DedicatedServer> server)
         else if (FMLEnvironment.dist == Dist.DEDICATED_SERVER) {
             return (T)this.serverSide.apply(server);
         }
+        else if (FMLEnvironment.dist == null) {
+            return (T)this.testSide.get();
+        }
         else {
             throw new IllegalArgumentException("THREE SIDES? WUT?");
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/SidedProvider.java]: This should update and say that custom sides are not supported

**Comment by liach** [comment]
By the way, is a consumer or try with resources better?

**Comment by mezz** [comment]
Consumer is better here because if something crashes during loading, we want it to just crash normally and it doesn't matter if the progress bar never completes. 
Try with resources would be better if we wanted to try to recover from crashes in some way and keep loading.

**Comment by mezz** [comment]
Hmm actually there may be some cases where try-with-resources is more flexible since you don't need to use a lambda, I'll try it out and see.

**Comment by liach** [comment]
try with resources does not need a catch phrase. It just closes an autocloseable on the try block exit. We can make the progress bar implement autocloseable, like how sponge deals with its cause stack manager.

**Comment by mezz** [comment]
Cleaned things up a bit and made `ProgressBar` implement `AutoCloseable`.
Added some basic unit tests.

**Pull request merged**